File: src/aobc/AlternateOneBitClock.go
package main

// To run the program:
// Install the Go and execute
// `go run AlternateOneBitClock.go`
// on prompt, terminal or PowerShell.

import (
	"fmt"
	"math/rand"
	"strconv"
)

// Generate a random number between 0 and 1,
// both included.
func randomBit() int {
	min := 0
	max := 1
	return (rand.Intn(max-min) + min)
}

// AlternateOneBitClock structure emulates
// the alternate one bit clock.
type AlternateOneBitClock struct {
	clk int
}

//Init willl initialize randomly clock variable clk
//with a value equal to zero or one.
func (a *AlternateOneBitClock) Init() {
	a.clk = randomBit()
	a.Print()
	a.Invariant()
}

// Print the clock value with ' -> ' sign.
func (a *AlternateOneBitClock) Print() {
	fmt.Print(a.clk, " -> ")
}

// Invariant asserts that the clock value has the desired value.
func (a *AlternateOneBitClock) Invariant() {
	if a.clk < 0 || a.clk > 1 {
		msg := "invariant not being held: value " +
			strconv.Itoa(a.clk) + " not expected\n"
		panic(msg)
	}
}

// Next state at each computation.
func (a *AlternateOneBitClock) Next() {
	for {
		if a.clk == 0 {
			a.clk = 1
		} else {
			a.clk = 0
		}
		a.Invariant()
		a.Print()

	}
}

func main() {
	ac := &AlternateOneBitClock{clk: -1}
	ac.Init()
	ac.Next()
}

File: src/aobc/AlternateOneBitClock.py
#!/usr/bin/env python3

# To run the program:
# Install the Python3 and execute
# `python3 AlternateOneBitClock.py`
# on prompt, terminal or PowerShell.

import random

class AlternateOneBitClock():
	'''Class to emulate the alternate one bit clock.
	'''	
	def __init__(self):
		'''Initialize randomly clock variable clk
		with a value equal to zero or one.
		'''
		self.clk = random.randint(0,1)
		self.Print()
		self.Invariant()

	def Print(self):
		'''Print the clock value with ' -> ' sign.
		'''
		print(self.clk, end=' -> ')		

	def Invariant(self):
		'''Assert the clock value has the desired value.
		'''
		assert self.clk == 0 or self.clk == 1

	def Next(self):
		'''Next state at each computation.
		'''
		while True:
			if self.clk == 0:
				self.clk = 1
			else:
				self.clk = 0
			
			self.Invariant()

			self.Print()

if __name__ == '__main__':
		ac = AlternateOneBitClock()
		ac.Next()
